streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java

```diff
@@ -131,11 +135,40 @@ public void setGlobalProcessorContext(final InternalProcessorContext globalProce
         }
 
         final Set<String> changelogTopics = new HashSet<>();
-        for (final StateStore stateStore : globalStateStores) {
+
+        long deadlineMs = NO_DEADLINE;
+        final List<StateStore> storesToInitialize = new LinkedList<>(globalStateStores);
+
+        while (!storesToInitialize.isEmpty()) {
+            // we remove and add back on failure to round-robin through all stores
+            final StateStore stateStore = storesToInitialize.remove(0);
             globalStoreNames.add(stateStore.name());
             final String sourceTopic = storeToChangelogTopic.get(stateStore.name());
             changelogTopics.add(sourceTopic);
-            stateStore.init(globalProcessorContext, stateStore);
+
+            try {
+                stateStore.init(globalProcessorContext, stateStore);
+                deadlineMs = NO_DEADLINE;
+            } catch (final RetryableErrorException retryableException) {
+                if (taskTimeoutMs == 0L) {
+                    throw new StreamsException(retryableException.getCause());
+                }
+
+                storesToInitialize.add(stateStore);
+
+                final long currentWallClockMs = time.milliseconds();
+                if (deadlineMs == NO_DEADLINE) {
+                    final long newDeadlineMs = currentWallClockMs + taskTimeoutMs;
+                    deadlineMs = newDeadlineMs < 0L ? Long.MAX_VALUE : newDeadlineMs;
+                } else if (currentWallClockMs > deadlineMs) {
+                    throw new TimeoutException(String.format(
+                        "Global task did not make progress to restore state within %d ms. Adjust `task.timeout.ms` if needed.",
+                        currentWallClockMs - deadlineMs + taskTimeoutMs
+                    ));
+                }
+
+                log.debug(retryableException.getMessage() + " Will retry. Remaining time in milliseconds: {}", deadlineMs - currentWallClockMs);
```

Comment:
```suggestion
                log.debug(String.format("Got an exception in store.init(). Will retry. Remaining time in milliseconds: %d",  deadlineMs - currentWallClockMs), retryableException);
```

Sorry, my prior feedback was ambiguous. This is what I meant.

Reply:
I understood what you meant, but I thought that because we actual do retry, it's better/cleaner/less-noisy to not log the full stack trace?

Reply:
Ah, good, at least it wasn't lack of clarity. It's your call. I've wasted quite a bit of time trying to track down root causes of things specifically because of stuff like this. It really doesn't seem like saving ten log lines is worth the extra hassle.